subworkflows/local/mtx_conversion.nf

```diff
@@ -1,62 +1,86 @@
 /* --    IMPORT LOCAL MODULES/SUBWORKFLOWS     -- */
-include { MTX_TO_H5AD   }             from '../../modules/local/mtx_to_h5ad.nf'
-include { CONCAT_H5AD   }             from '../../modules/local/concat_h5ad.nf'
-include { MTX_TO_SEURAT }             from '../../modules/local/mtx_to_seurat.nf'
+include { MTX_TO_H5AD           } from '../../modules/local/mtx_to_h5ad'
+include { CONCAT_H5AD           } from '../../modules/local/concat_h5ad.nf'
+include { ANNDATAR_CONVERT      } from '../../modules/local/anndatar_convert'
+include { EMPTY_DROPLET_REMOVAL } from '../../subworkflows/local/emptydrops_removal'
 
 workflow MTX_CONVERSION {
 
     take:
     mtx_matrices
-    samplesheet
     txp2gene
     star_index
+    samplesheet
 
     main:
         ch_versions = Channel.empty()
-
-        // Cellranger module output contains too many files which cause path collisions, we filter to the ones we need.
-        // Keeping backwards compatibility with cellranger-arc.
-        // TODO: Adapt cellranger-arc subworkflow like cellranger to remove this snippet here.
-        if (params.aligner in [ 'cellrangerarc' ]) {
-            mtx_matrices = mtx_matrices.map { meta, mtx_files ->
-                    [ meta, mtx_files.findAll { it.toString().contains("filtered_feature_bc_matrix") } ]
-                }
-                .filter { meta, mtx_files -> mtx_files } // Remove any that are missing the relevant files
-        }
+        ch_h5ads    = Channel.empty()
 
         //
-        // Convert matrix to h5ad
+        // MODULE: Convert matrices to h5ad
         //
         MTX_TO_H5AD (
             mtx_matrices,
             txp2gene,
             star_index
         )
+        ch_versions = ch_versions.mix(MTX_TO_H5AD.out.versions.first())
+
+        // fix channel size when kallisto non-standard workflow
+        if (params.aligner == 'kallisto' && !(params.kb_workflow == 'standard')) {
+            ch_h5ads =
+            MTX_TO_H5AD.out.h5ad
+                .transpose()
+                .map { meta, h5ad ->
+                    def meta_clone = meta.clone()
+                    def spc_prefix = h5ad.toString().contains('unspliced') ? 'un' : ''
+
+                    meta_clone["input_type"] = "${meta.input_type}_${spc_prefix}spliced"
+
+                    [ meta_clone, h5ad ]
+                }
+        } else {
+            ch_h5ads = MTX_TO_H5AD.out.h5ad
+        }
+
+        //
+        // SUBWORKFLOW: Run cellbender emptydrops filter
+        //
+        if ( !params.skip_emptydrops && !(params.aligner in ['cellrangerarc']) ) {
+
+            // emptydrops should only run on the raw matrices thus, filter-out the filtered result of the aligners that can produce it
+            EMPTY_DROPLET_REMOVAL (
+                ch_h5ads.filter { meta, mtx_files -> meta.input_type.contains('raw') }
+            )
+            ch_h5ads = ch_h5ads.mix( EMPTY_DROPLET_REMOVAL.out.h5ad )
```

Comment:
I would put that ouside the `mtx_conversion` workflow. I think the mtx conversion should just take care of converting mtx -> h5ad. 

Droplet removal can be a separate unit that works on h5ad files then.

Reply:
Hi @grst ,
I would think the same, the problem here, is that the module for emptydrops call requires and h5ad already, which requires the conversion :/

https://github.com/nf-core/scrnaseq/blob/standardize-conversion-workflow/modules/nf-core/cellbender/removebackground/main.nf

Reply:
yes, that's clear, I'd just put it **after** the mtx conversion workflow, not **inside** ;)

Reply:
I am trying to do like this.

The question here would be, in case this happens, I am not seeing how we could easily make the emptydrops filtered file be converted and also be included for the results concatenation, as this would happen before the emptydrops is called.

Maybe we can have a quick chat to better discuss this part specifically.

Reply:
`concat_h5ad` could also be a standalone module, no? 

We could even have a `convert_to_rds + concat_h5ad` subworkflow that is called multiple times.

Reply:
happy to chat or have a quick call if it helps :)